### ISSUE TITLE
Safe enum - turns on FS104 by default

### ISSUE THREAD

**Issue by jwosty** [fsharp/fslang-suggestions#1052]
# Title of Suggestion

I propose we have a new way to define enums which makes from integers safer, and disables FS104 ("enums may take values outside known cases") for types defined using it. The syntax is arbitrary and up for discussion, but to spark conversion, I propose the following syntax:

    [<SafeEnum>]
    type Polygon = Triangle = 3 | Square = 4 | Pentagon = 5

And the following code would either produce a compiler error, or produce a warning and possibly fail at runtime after checking if the integer is a valid enum case:

    enum<Polygon> 5
    // Could either be a compiler error, saying something akin to "Cannot cast int to Polygon"
    // Or, could warn "Casting from an integer to Polygon is an unsafe operation and could fail at runtime"

Then, the following code would compile without warning:

    let getPolygonName polygon =
        match polygon with
        | Polygon.Triangle -> "triangle"
        | Polygon.Square -> "square"
        | Polygon.Pentagon -> "pentagon"

And incomplete pattern matches would behave as normal for enums.

Additionally, a "safe enum" with FlagsAttribute would elicit a compiler error.

## Pros and Cons 

Advantages:
    * Provides an version of enums that are *almost* as type-safe as DU's (as disabling FS104 makes the assumption that the code providing the enum value isn't doing anything evil, like casting from integers)

    * It still wouldn't be a perfect guard against unsafe enum casting (i.e. could bypass with reflection, or incoming C# code)
        * Counterpoint - it would reduce the surface area of possible mistakes when used

## Extra information

Estimated cost (XS, S, M, L, XL, XXL): S

Related suggestions: #652

## Affidavit (please submit!)

Please tick this by placing a cross in the box:
* [x] This is not a question (e.g. like one you might ask on [stackoverflow](http://stackoverflow.com)) and I have searched stackoverflow for discussions of this issue
* [x] I have [searched both open and closed suggestions on this site](http://github.com/fsharp/fslang-suggestions/issues) and believe this is not a duplicate
* [ ] This is not something which has obviously "already been decided" in previous versions of F#.  If you're questioning a fundamental design decision that has obviously already been taken (e.g. "Make F# untyped") then please don't submit it.

Please tick all that apply:
* [x] This is not a breaking change to the F# language design
* [x] I or my company would be willing to help implement and/or test this

## For Readers

If you would like to see this issue implemented, please click the :+1: emoji on this issue. These counts are used to generally order the suggestions by engagement.

**Comment by Happypig375** [comment]
What happens at interop boundaries?

**Comment by BentTranberg** [comment]
I can't see how this would be helpful in any way. We already have DUs to solve this kind of thing. Better to just leave enums to exactly what they are, no more no less - a .NET and C#/VB thing.

The same thing can easily be accomplished with a DU and a function that return numeric values. I don't see what other use cases this feature would cover.

```
type Polygon = Triangle | Square | Pentagon

let valOfPolygon = function Triangle -> 3 | Square -> 4 | Pentagon -> 5
```

**Comment by jwosty** [comment]
> What happens at interop boundaries?

That's a good question, honestly. I can't think of a good solution yet.

> The same thing can easily be accomplished with a DU and a function that return numeric values. I don't see what other use cases this feature would cover.

You're right; it serves the same purpose. The main nicety about it is that you don't have to write that boilerplate; you can just freely cast back and forth. It would be purely for convenience to not have to write that function yourself.

**Comment by Swoorup** [comment]
Isn't there a compiler flag to disable the current warning? If you are sure code isn't used across boundaries with other .net languages?

**Comment by BentTranberg** [comment]
I just want to point out some details that I noticed when reading the suggestion, of which some caused a bit of confusion. Also adding a bit to the discussion. Still not convinced this is a good idea, but willing to explore.

The title is "Safe enum - turns on FS104 by default" and then it says "and disables FS104". I'm confused.

FS104 - The correct error code is FS0104. Use it, so that searching it will match this issue.

The heading "Title of Suggestion" should be the same as the issue title.

`Pentagon = 5`, yet `enum<Polygon> 5` gives a warning or an error? Maybe use another number for this example.

> // Could either be a compiler error, saying something akin to "Cannot cast int to Polygon"

"Cannot cast int to Polygon" - taken very literally it's wrong, but okay, it did say "something akin to". Maybe "Cannot cast 6 to Polygon" or "Enum value 6 is not defined".

> // Or, could warn "Casting from an integer to Polygon is an unsafe operation and could fail at runtime"

This warning is troublesome. It'd be like warning that a division could fail at runtime. It is only acceptable if it is meant as a way to warn the developer to use another way to convert a "safe enum". (Would happen after you add `[<SafeEnum>]` to an "old" enum, so that would be ok, to get a warning it might blow up.) Another way is to have a cast/function that returns a "safe enum" option.

`safeenum<Polygon> 5`

I found another related suggestion: #822

**Comment by jwosty** [comment]
> 
> 
> Isn't there a compiler flag to disable the current warning? If you are sure code isn't used across boundaries with other .net languages?

Yes, and that would be the current workaround. Disabling the warning doesn't take away the really easy ways to cast ints to enums, however. I.e. it wouldn't stop you from doing:
```fsharp
#nowarn "0104"
let value = enum<Polygon> 10
```

This safe enum mechanism would block off the simple routes like that (there's always reflection, sure, but if you're using reflection, you're signing up to be okay with potentially breaking plenty of core features).

>
> I just want to point out some details that I noticed when reading the suggestion, of which some caused a bit of confusion.

Thank you for catching those typos. They should be fixed now.

>
> Still not convinced this is a good idea, but willing to explore.

Fair enough. I appreciate the indulgence. :)

>> // Could either be a compiler error, saying something akin to "Cannot cast int to Polygon"
> "Cannot cast int to Polygon" - taken very literally it's wrong, but okay, it did say "something akin to". Maybe "Cannot cast 6 to Polygon" or "Enum value 6 of Polygon is not defined".

I see your point. ``Enum value 6 of Polygon is not defined`` is pretty good actually. As another option, ``enum<Polygon> (6) is not defined`` might also make some sense as a parallel of the warning text in FS0104 (which would say things like ``Enums may take values outside known cases. For example, the value 'enum<Polygon> (6)' may indicate a case not covered by the pattern(s).``



>>    // Or, could warn "Casting from an integer to Polygon is an unsafe operation and could fail at runtime"

> This warning is troublesome. It'd be like warning that a division could fail at runtime. It is only acceptable if it is meant as a way to warn the developer to use another way to convert a "safe enum". (Would happen after you add [<SafeEnum>] to an "old" enum, so that would be ok, to get a warning it might blow up.) Another way is to have a cast/function that returns a "safe enum" option.

Perhaps we could have a ``tryParseEnum<'T> : int -> 'T option`` function -- if the input is a valid enum case value, then it would return a Some of the case; if not, None. For consistency maybe it should also do the same thing for normal enums (though I'm not convinced of this -- it could be a function really only meant for safe enums). Then, we could issue a compiler *error* instead of warning for usages of ``enum<'T>`` on safe enums.

**Comment by charlesroddie** [comment]
This an attribute attached to a .net enum that indicates that it should be treated more safely?

> Perhaps we could have a `tryParseEnum<'T> : int -> 'T option` function

If `enum<'T>:int->'T` is blocked we would definitely need a `tryParseEnum`. The syntax involves numbers, so the numbers need to be usable. (E.g. an advantage of numbers is serialization, and then we need int -> enum in some form to deserialize.)

Are the enum cases known at runtime to allow a `tryParseEnum` to be defined?

Or is the idea to have a special form of struct DU with the numbers in the tag determined?

**Comment by BentTranberg** [comment]
My understanding is that this suggestion is about an alternative way to treat .NET enums in the F# compiler, thus still be interoperable with C#/.NET, and not to introduce a new F# type. (Not that it should hinder us discussing alternatives of course.)

> tryParseEnum

That's what I meant, but you made it clear, not least with a perfect name.
